src/blog/2018-09-24/mutation-switching.pug

```diff
@@ -0,0 +1,116 @@
+                        a >= b  
+                    case Some("1") =>  
+                        a < b  
+                    case Some("2") =>  
+                        a == b  
+                    case _ =>  
+                        a > b  
+                }
+            }  
+        }
+        ```
+        All possible mutations are implemented in the pattern match with an identifier so we can turn on specific mutations 
+        and a default case so the code will work normally one none of the mutants are activated. Now the code base only needs to be compiled once, 
+        but because of the increased size of the code base the compilation time will be a bit longer but that does not way up 
+        against the full recompilation as before. If the compilation time for the code base would be 15 seconds we will still 
+        gain 15 seconds which is already a significant performance booster. So we gained a lot of performance without losing any 
+        flexibility because of the ability to turn mutations on and off using the environment variable. 
+        
+        ## Top statements
+        With mutation switching, we reduced compilation times by a lot but soon noticed there was one more problem to tackle. 
+        Previously a very trivial example was used. With this statement, it was easy to mutate using mutation switching. 
+        But what if the statements get more complex? How can complex code be avoided? Because the more complex the code is the higher change compilation errors occur. 
+        For example
+        ```scala
+        def isEven(number: Int): Boolean = number % 2 == 0
+        def isOdd(number: Int): Boolean = !isEven(number)
+
+        val numbers: List[Int] = 1 to 100 toList
+
+        def filterOddAnd(specialNumber: Int): List[Int] = {
+          numbers
+            .filter(isOdd)
+            .filterNot(_.equals(specialNumber))
+        }
+        ```
+        With this code base `filter` and `filterNot` could be mutated to there counterparts. 
+        This would give us the following code base if we implement the pattern match at the direct position. 
+        ```scala
+        def isEven(number: Int): Boolean = number % 2 == 0
+        def isOdd(number: Int): Boolean = !isEven(number)
+
+        val numbers: List[Int] = 1 to 100 toList
+
+        def filterOddAnd(specialNumber: Int): List[Int] = {
+          numbers.(sys.env.get("ACTIVE_MUTATION") match {
+            case Some("1") => .filter(isOdd)
+            case _ => .filterNot(isOdd)
+          }).
+          (sys.env.get("ACTIVE_MUTATION") match {
+            case Some("1") =>  .filterNot(_.equals(specialNumber))
+            case _ =>  .filter(_.equals(specialNumber))
+          })
+        }
+        ```
+        As you can see it's already a lot more complicated and the code doesn't even compile! 
+        To reduce a lot of the complexity and make the code actually compile we need to take a deeper look in the abstract syntax tree 
```

Comment:
I would leave out the "complexity" part. It makes the text opinionated, let the reader decide for itself if it is complex or not. The most important take away is that the code doesn't compile, right?